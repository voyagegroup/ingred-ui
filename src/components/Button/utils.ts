import { Theme } from "../../themes";
import { ButtonColor, ButtonColorStyle } from "./Button";

export const getContainerColorStyles = (
  theme: Theme,
): { [P in ButtonColor]: ButtonColorStyle } => ({
  primary: {
    normal: {
      background: theme.palette.primary.main,
      color: theme.palette.text.white,
      boxShadow: theme.shadow["3dShadowPrimary"],
      border: `1px solid ${theme.palette.primary.dark}`,
    },
    hover: {
      background: theme.palette.primary.dark,
      border: "none",
    },
    active: {
      background: theme.palette.primary.dark,
      boxShadow: theme.shadow["3dShadowActive"],
      border: "none",
    },
  },
  basicLight: {
    normal: {
      background: theme.palette.background.default,
      color: theme.palette.black,
      boxShadow: theme.shadow["3dShadowBasic"],
      border: `1px solid ${theme.palette.divider}`,
    },
    hover: {
      background: theme.palette.gray.highlight,
      border: `1px solid ${theme.palette.divider}`,
    },
    active: {
      background: theme.palette.gray.highlight,
      boxShadow: theme.shadow["3dShadowActive"],
      border: `1px solid ${theme.palette.divider}`,
    },
  },
  basicDark: {
    normal: {
      background: theme.palette.basicDark.main,
      color: theme.palette.black,
      boxShadow: theme.shadow["3dShadowBasic"],
      border: `1px solid ${theme.palette.basicDark.deepDark}`,
    },
    hover: {
      background: theme.palette.basicDark.dark,
      border: "none",
    },
    active: {
      background: theme.palette.basicDark.dark,
      boxShadow: theme.shadow["3dShadowBasicActive"],
      border: "none",
    },
  },
  danger: {
    normal: {
      background: theme.palette.danger.main,
      color: theme.palette.text.white,
      boxShadow: theme.shadow["3dShadowDanger"],
      border: `1px solid ${theme.palette.danger.dark}`,
    },
    hover: {
      background: theme.palette.danger.dark,
      border: "none",
    },
    active: {
      background: theme.palette.danger.dark,
      boxShadow: theme.shadow["3dShadowActive"],
      border: "none",
    },
  },
  clear: {
    normal: {
      background: "none",
      color: theme.palette.black,
      boxShadow: "none",
      border: "none",
    },
    hover: {
      background: theme.palette.gray.light,
      border: "none",
    },
    active: {
      background: theme.palette.gray.main,
      boxShadow: "none",
      border: "none",
    },
  },
  primaryPale: {
    normal: {
      background: theme.palette.primaryPale.main,
      color: theme.palette.primaryPale.dark,
      boxShadow: theme.shadow["3dShadowPrimaryPale"],
      border: `1px solid ${theme.palette.primaryPale.light}`,
    },
    hover: {
      background: theme.palette.primaryPale.highlight,
      border: "none",
    },
    active: {
      background: theme.palette.primaryPale.highlight,
      boxShadow: theme.shadow["3dShadowPrimaryPaleActive"],
      border: "none",
    },
  },
});
