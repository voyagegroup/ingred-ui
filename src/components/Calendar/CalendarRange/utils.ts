import dayjs, { Dayjs } from "dayjs";

const isStart = (startDate: Dayjs | null, month: Dayjs, day: number) =>
  startDate?.format("YYYY-MM-DD") ===
  dayjs(new Date(month.year(), month.month(), day)).format("YYYY-MM-DD");

const isEnd = (endDate: Dayjs | null, month: Dayjs, day: number) =>
  endDate?.format("YYYY-MM-DD") ===
  dayjs(new Date(month.year(), month.month(), day)).format("YYYY-MM-DD");

const isBetween = (
  startDate: Dayjs | null,
  endDate: Dayjs | null,
  month: Dayjs,
  day: number,
) =>
  (startDate &&
    endDate &&
    dayjs(new Date(month.year(), month.month(), day)).isAfter(
      startDate.format("YYYY-MM-DD"),
      "day",
    ) &&
    dayjs(new Date(month.year(), month.month(), day)).isBefore(
      endDate.format("YYYY-MM-DD"),
      "day",
    )) ??
  false;

export const getDayState = (
  startDate: Dayjs | null,
  endDate: Dayjs | null,
  month: Dayjs,
  day: number,
) => {
  if (isStart(startDate, month, day)) {
    return "start";
  }
  if (isEnd(endDate, month, day)) {
    return "end";
  }
  if (isBetween(startDate, endDate, month, day)) {
    return "between";
  }
  return "none";
};
