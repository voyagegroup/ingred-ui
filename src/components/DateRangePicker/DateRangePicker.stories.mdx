import { Meta, Story, ArgsTable, Canvas } from "@storybook/addon-docs/blocks";
import moment from "moment";
import DateRangePicker from "./DateRangePicker";
import "react-dates/lib/css/_datepicker.css";

<Meta title="Components/Inputs/DateRangePicker" component={DateRangePicker} />

# DateRangePicker

The wrapper of [react-dates](https://github.com/airbnb/react-dates).

For more detail props, please see [it](https://github.com/airbnb/react-dates#daterangepicker).

<ArgsTable of={DateRangePicker} />

## When the display is strange

Please import css from `react-dates`.

```tsx
import "react-dates/lib/css/_datepicker.css";
```

## Samples

### Basic

<Canvas>
  <Story name="Basic">
    {() => {
      {
        /* MEMO: To be unaffected by "Localize" story. */
      }
      moment.locale("en");
      const [startDate, setStartDate] = React.useState(moment().set("date", 1));
      const [endDate, setEndDate] = React.useState(moment());
      const handleChangeDates = ({ startDate, endDate }) => {
        setStartDate(startDate);
        setEndDate(endDate);
      };
      return (
        <div style={{ height: "400px" }}>
          <DateRangePicker
            startDate={startDate}
            endDate={endDate}
            onDatesChange={handleChangeDates}
          />
        </div>
      );
    }}
  </Story>
</Canvas>

### Error

<Canvas>
  <Story name="Error">
    {() => (
      <DateRangePicker
        startDate={moment().set("date", 1)}
        endDate={moment()}
        error={true}
      />
    )}
  </Story>
</Canvas>

### Localize

<Canvas>
  <Story name="Localize">
    {() => {
      moment.locale("ja", {
        weekdaysShort: ["日", "月", "火", "水", "木", "金", "土"],
      });
      const renderMonthText = (day) => day.format("YYYY年M月");
      const displayFormat = () => "YYYY/MM/DD";
      const [startDate, setStartDate] = React.useState(moment().set("date", 1));
      const [endDate, setEndDate] = React.useState(moment());
      const handleChangeDates = ({ startDate, endDate }) => {
        setStartDate(startDate);
        setEndDate(endDate);
      };
      return (
        <div style={{ height: "400px" }}>
          <DateRangePicker
            startDate={startDate}
            endDate={endDate}
            displayFormat={displayFormat}
            renderMonthText={renderMonthText}
            onDatesChange={handleChangeDates}
          />
        </div>
      );
    }}
  </Story>
</Canvas>
