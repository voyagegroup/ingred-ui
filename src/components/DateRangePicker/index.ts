export { default } from "./DateRangePicker";
export type { DateRangePickerProps, DateRange } from "./DateRangePicker";
