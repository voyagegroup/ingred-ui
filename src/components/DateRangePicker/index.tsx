export { default } from "./DateRangePicker";
export type { DateRangePickerProps } from "./DateRangePicker";
