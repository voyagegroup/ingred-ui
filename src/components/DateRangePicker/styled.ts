import styled from "styled-components";

export const Container = styled.div<{ error: boolean }>`
  /* Overriding styles */
  /* DateRangePicker */
  .DateRangePicker {
    display: inline-flex;
  }

  /* DateRangePickerInput */
  .DateRangePickerInput {
    display: block;
    border: 1px solid
      ${({ error, theme }) =>
        error ? theme.palette.danger.main : theme.palette.divider};
    border-radius: ${({ theme }) => theme.radius}px;
    background-color: ${({ theme }) => theme.palette.background.default};
    overflow: hidden;
  }
  .DateInput {
    width: 96px;
    background-color: transparent;
  }
  .DateInput_input {
    padding: 10px 11px 8px;
    line-height: 20px;
    font-size: 14px;
    color: ${({ error }) =>
      error
        ? ({ theme }) => theme.palette.danger.main
        : ({ theme }) => theme.palette.black};
    background-color: transparent;
    font-family: "Noto Sans JP", "Hiragino Kaku Gothic ProN", "Proxima Nova",
      Verdana, "游ゴシック", YuGothic, Meiryo, sans-serif;
  }
  .DateInput_input__focused {
    border-bottom-color: ${({ theme }) => theme.palette.primary.main};
  }
  /* DayPickerRangeController */
  .DayPicker__withBorder {
    box-shadow: 0px 0px 16px #041c3315;
    border-radius: ${({ theme }) => theme.radius}px;
  }
  .DayPicker_weekHeader {
    color: ${({ theme }) => theme.palette.text.secondary};
  }
  td {
    position: relative;
  }
  td:nth-child(7n):not(.CalendarDay__selected_start) {
    &:before {
      left: 0;
      border-top-right-radius: 11px;
      border-bottom-right-radius: 11px;
    }
    &:after {
      left: 0;
    }
  }
  td:nth-child(7n).CalendarDay__selected_span,
  td:nth-child(7n).CalendarDay__hovered_span {
    &:before {
      width: calc((41px - 22px) / 2 + 22px);
    }
    &:before {
      width: calc((41px - 22px) / 2 + 22px);
    }
  }
  td:nth-child(7n + 1):not(.CalendarDay__selected_end) {
    &:before {
      right: 0;
      border-top-left-radius: 11px;
      border-bottom-left-radius: 11px;
    }
    &:after {
      right: 0;
    }
  }
  td:nth-child(7n + 1).CalendarDay__selected_span,
  td:nth-child(7n + 1).CalendarDay__hovered_span {
    &:before {
      width: calc((41px - 22px) / 2 + 22px);
    }
  }
  .CalendarMonth_caption {
    font-size: 16px;
    padding-bottom: 50px;
  }
  .CalendarDay__blocked_out_of_range {
    border: 0;
  }
  .CalendarDay__default {
    border: 0;
    font-size: 12px;
    background: transparent;
    z-index: 1;
    &:hover {
      background: inherit;
      border: 0;
      &:before {
        content: "";
        display: block;
        position: absolute;
        top: 50%;
        /* MEMO: Use !important to prevent overwriting by 'td:nth-child(7n):before' selector */
        left: 50% !important;
        transform: translate(calc(-50% - 0.5px), -50%);
        width: 22px;
        height: 22px;
        border-radius: 50%;
        z-index: -1;
        background: ${({ theme }) => theme.palette.gray.main};
      }
    }
  }
  .CalendarDay__hovered_span {
    background: transparent;
    color: ${({ theme }) => theme.palette.black};
    z-index: 1;
    &:hover {
      background: inherit;
      border: 0;
      color: inherit;
    }
    &:before {
      content: "";
      display: block;
      position: absolute;
      top: 50%;
      /* MEMO: 'left: 0;' is written by 'td:nth-child' */
      transform: translateY(-50%);
      width: 100%;
      height: 22px;
      background: ${({ theme }) => theme.palette.primary.highlight};
      z-index: -1;
    }
  }
  .CalendarDay__selected_span {
    background: transparent;
    color: ${({ theme }) => theme.palette.black};
    z-index: 1;
    &:hover {
      background: inherit;
      border: 0;
      color: inherit;
      &:before {
        background: ${({ theme }) => theme.palette.background.hint};
      }
    }
    &:before {
      content: "";
      display: block;
      position: absolute;
      top: 50%;
      /* MEMO: 'left: 0;' is written by 'td:nth-child' */
      transform: translateY(-50%);
      width: 100%;
      height: 22px;
      background: ${({ theme }) => theme.palette.primary.highlight};
      z-index: -1;
    }
  }
  .CalendarDay__outside {
    color: ${({ theme }) => theme.palette.text.secondary};
  }
  .CalendarDay__selected {
    background: transparent;
    color: ${({ theme }) => theme.palette.white};
    z-index: 1;
    &:hover {
      background: inherit;
      border: 0;
      color: ${({ theme }) => theme.palette.white};
      &:before {
        background: ${({ theme }) => theme.palette.primary.dark};
      }
      &::after {
        background: ${({ theme }) => theme.palette.background.hint};
      }
    }
    &:before {
      content: "";
      display: block;
      position: absolute;
      top: 50%;
      /* MEMO: Use !important to prevent overwriting by 'td:nth-child(7n):before' selector */
      left: 50% !important;
      transform: translate(calc(-50% - 0.5px), -50%);
      width: 22px;
      height: 22px;
      background: ${({ theme }) => theme.palette.primary.main};
      border-radius: 50%;
      z-index: -1;
    }
  }
  .CalendarDay__selected_start {
    &::after {
      content: "";
      display: block;
      position: absolute;
      top: 50%;
      left: 50%;
      transform: translateY(-50%);
      width: 50%;
      height: 22px;
      background: ${({ theme }) => theme.palette.primary.highlight};
      z-index: -2;
    }
  }
  .CalendarDay__selected_end {
    &::after {
      content: "";
      display: block;
      position: absolute;
      top: 50%;
      right: 50%;
      transform: translateY(-50%);
      width: 50%;
      height: 22px;
      background: ${({ theme }) => theme.palette.primary.highlight};
      z-index: -2;
    }
  }
`;

export const CustomArrowIcon = styled.div`
  width: 8px;
  height: 1px;
  background-color: ${({ theme }) => theme.palette.text.hint};
`;

export const BaseNavIcon = styled.div`
  position: absolute;
  top: 22px;
  border-radius: 50%;
  background-color: ${({ theme }) => theme.palette.gray.light};
`;

export const NavPrev = styled(BaseNavIcon)`
  left: 30px;
`;

export const NavNext = styled(BaseNavIcon)`
  right: 30px;
`;
