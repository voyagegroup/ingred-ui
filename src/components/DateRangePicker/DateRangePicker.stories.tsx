import { StoryObj } from "@storybook/react";
import DateRangePicker, { DateRangePickerProps } from "./DateRangePicker";
import dayjs from "dayjs";
import React, { useState } from "react";

export default {
  title: "Components/Inputs/DateRangePicker",
  component: DateRangePicker,
  args: {
    format: "YYYY-MM-DD",
  },
};

export const Example: StoryObj<DateRangePickerProps> = {
  render: (args) => {
    const [date, setDate] = useState({
      startDate: dayjs(),
      endDate: dayjs().add(1, "week"),
    });

    return (
      <DateRangePicker
        {...args}
        startDate={date.startDate}
        endDate={date.endDate}
        onDatesChange={setDate}
      />
    );
  },
};

export const WithActions: StoryObj<DateRangePickerProps> = {
  render: (args) => {
    const [date, setDate] = useState({
      startDate: dayjs(),
      endDate: dayjs().add(1, "week"),
    });
    const actions = [
      {
        text: "明日",
        onClick: () => {
          setDate({
            startDate: dayjs(),
            endDate: dayjs().add(1, "day"),
          });
        },
      },
      {
        text: "来週",
        onClick: () => {
          setDate({
            startDate: dayjs(),
            endDate: dayjs().add(1, "week"),
          });
        },
      },
      {
        text: "先月",
        onClick: () => {
          setDate({
            startDate: dayjs().subtract(1, "month"),
            endDate: dayjs(),
          });
        },
      },
    ];

    return (
      <DateRangePicker
        {...args}
        actions={actions}
        startDate={date.startDate}
        endDate={date.endDate}
        onDatesChange={setDate}
      />
    );
  },
};

export const WithActionsWithDefaultClickAction: StoryObj<DateRangePickerProps> =
  {
    render: (args) => {
      const [date, setDate] = useState({
        startDate: dayjs(),
        endDate: dayjs().add(1, "week"),
      });
      const [clickAction, setClickAction] = useState("来週");
      const actions = [
        {
          text: "明日",
          onClick: () => {
            setDate({
              startDate: dayjs(),
              endDate: dayjs().add(1, "day"),
            });
          },
        },
        {
          text: "来週",
          onClick: () => {
            setDate({
              startDate: dayjs(),
              endDate: dayjs().add(1, "week"),
            });
          },
        },
        {
          text: "先月",
          onClick: () => {
            setDate({
              startDate: dayjs().subtract(1, "month"),
              endDate: dayjs(),
            });
          },
        },
      ];

      const handleDateChange = (dates: {
        startDate: dayjs.Dayjs;
        endDate: dayjs.Dayjs;
      }) => {
        setDate(dates);
        setClickAction("");
      };

      return (
        <DateRangePicker
          {...args}
          actions={actions}
          startDate={date.startDate}
          endDate={date.endDate}
          defaultClickAction={clickAction}
          onClickAction={(action) => setClickAction(action.text)}
          onDatesChange={handleDateChange}
        />
      );
    },
  };

export const Error: StoryObj<DateRangePickerProps> = {
  ...Example,
  args: {
    errorText: "エラー",
  },
};

export const Disabled: StoryObj<DateRangePickerProps> = {
  ...Example,
  args: {
    disabled: true,
  },
};

export const IsOutsideRange: StoryObj<DateRangePickerProps> = {
  render: (args) => {
    const [date, setDate] = useState({
      startDate: dayjs(),
      endDate: dayjs().add(1, "week"),
    });

    const isOutsideRange = (day: dayjs.Dayjs) =>
      day.isBefore(dayjs().subtract(1, "day"));

    return (
      <DateRangePicker
        {...args}
        startDate={date.startDate}
        endDate={date.endDate}
        isOutsideRange={isOutsideRange}
        onDatesChange={setDate}
      />
    );
  },
};
