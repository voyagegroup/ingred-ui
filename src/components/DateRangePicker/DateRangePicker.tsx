import * as React from "react";
import * as Styled from "./styled";
import "react-dates/initialize";
import moment from "moment";
import {
  FocusedInputShape,
  DateRangePicker as OriginalDateRangePicker,
  DateRangePickerShape,
} from "react-dates";
import Icon from "../Icon";
import Spacer from "../Spacer";
import { useLocaleProps } from "../../hooks/useLocaleProps";

function isOutsideRange() {
  return false;
}

export type DateRangePickerProps = Partial<DateRangePickerShape> & {
  startDate: moment.Moment | null;
  endDate: moment.Moment | null;
  onDatesChange: (arg: {
    startDate: moment.Moment | null;
    endDate: moment.Moment | null;
  }) => void;
  error?: boolean;
};

const DateRangePicker = React.forwardRef<HTMLDivElement, DateRangePickerProps>(
  (inProps, ref) => {
    const props = useLocaleProps({ props: inProps, name: "DateRangePicker" });
    const { startDate, endDate, error = false, ...rest } = props;
    const [focusedInput, setFocusedInput] =
      React.useState<FocusedInputShape | null>(null);

    return (
      <Styled.Container ref={ref} error={error}>
        <OriginalDateRangePicker
          startDatePlaceholderText="FROM"
          endDatePlaceholderText="TO"
          isOutsideRange={isOutsideRange}
          customArrowIcon={<Styled.CustomArrowIcon />}
          numberOfMonths={2}
          enableOutsideDays={true}
          daySize={41}
          weekDayFormat="ddd"
          hideKeyboardShortcutsPanel={true}
          startDateId="startDate"
          endDateId="endDate"
          navPrev={
            <Styled.NavPrev>
              <Spacer p={0.5}>
                <Icon name="arrow_left" size="md" />
              </Spacer>
            </Styled.NavPrev>
          }
          navNext={
            <Styled.NavNext>
              <Spacer p={0.5}>
                <Icon name="arrow_right" size="md" />
              </Spacer>
            </Styled.NavNext>
          }
          {...rest}
          startDate={startDate}
          endDate={endDate}
          focusedInput={focusedInput}
          onFocusChange={setFocusedInput}
        />
      </Styled.Container>
    );
  },
);

export default DateRangePicker;
