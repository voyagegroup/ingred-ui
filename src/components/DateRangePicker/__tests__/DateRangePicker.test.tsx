import * as React from "react";
import "@testing-library/jest-dom/extend-expect";
import { cleanup } from "@testing-library/react";
import DateRangePicker from "..";
import { renderWithThemeProvider } from "../../../utils/renderWithThemeProvider";

describe("DateRangePicker component testing", () => {
  afterEach(cleanup);

  test("DateRangePicker", () => {
    const { asFragment } = renderWithThemeProvider(
      <DateRangePicker
        startDate={null}
        endDate={null}
        onDatesChange={jest.fn()}
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
