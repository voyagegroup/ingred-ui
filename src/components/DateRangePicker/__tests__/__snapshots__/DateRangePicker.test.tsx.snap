// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateRangePicker component testing DateRangePicker 1`] = `
<DocumentFragment>
  <div
    class="sc-bdvvtL gCpgqx"
  >
    <div
      class="DateRangePicker DateRangePicker_1"
    >
      <div>
        <div
          class="DateRangePickerInput DateRangePickerInput_1 DateRangePickerInput__withBorder DateRangePickerInput__withBorder_2"
        >
          <div
            class="DateInput DateInput_1"
          >
            <input
              aria-describedby="DateInput__screen-reader-message-startDate"
              aria-label="FROM"
              autocomplete="off"
              class="DateInput_input DateInput_input_1"
              id="startDate"
              name="startDate"
              placeholder="FROM"
              type="text"
              value=""
            />
            <p
              class="DateInput_screenReaderMessage DateInput_screenReaderMessage_1"
              id="DateInput__screen-reader-message-startDate"
            >
              Navigate forward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.
            </p>
          </div>
          <div
            aria-hidden="true"
            class="DateRangePickerInput_arrow DateRangePickerInput_arrow_1"
            role="presentation"
          >
            <div
              class="sc-gsDKAQ bTdIb"
            />
          </div>
          <div
            class="DateInput DateInput_1"
          >
            <input
              aria-describedby="DateInput__screen-reader-message-endDate"
              aria-label="TO"
              autocomplete="off"
              class="DateInput_input DateInput_input_1"
              id="endDate"
              name="endDate"
              placeholder="TO"
              type="text"
              value=""
            />
            <p
              class="DateInput_screenReaderMessage DateInput_screenReaderMessage_1"
              id="DateInput__screen-reader-message-endDate"
            >
              Navigate backward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
